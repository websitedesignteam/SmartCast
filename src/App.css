.App {
  position: relative;
  min-height: 100vh;
}

.logo {
  margin: auto;
  width: 20rem;
}

body {
  font-family: "Cairo";
}

button {
  cursor: pointer;
  background-color: #dfdfdf;
}

button:hover:not([disabled]) {
	background-color: #d3d3d3;
}