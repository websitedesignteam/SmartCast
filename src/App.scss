@import "styles/variables";

.App {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  margin: 0;
}

.App-content {
//   padding: 1.5rem;
  min-height: 100vh;
}

.App-blur {
  position: absolute;
  top: 4rem;
  bottom: 0;
  left: 0;
  z-index: 1;
  backdrop-filter: blur(0.1rem);
  padding: 0;
  min-width: 100vw;
  // @media (orientation: landscape) {
  //   min-height: 170vh;
  // }
  // @media screen and (min-width: 400px) and (orientation: landscape) {
  //   min-height: 80vh;
  // }
}

.App-background {
  position: absolute;
  top: 4rem;
  bottom: 0;
  left: 0;
  z-index: 1;
  padding: 0;
  min-width: 100vw;
  background-image: url("/assets/background.png");
  background-color: $light-gray;
  min-height: 80vh;
  @media (orientation: landscape) {
    min-height: 170vh;
  }
  @media screen and (min-width: 400px) and (orientation: landscape) {
    min-height: 80vh;
  }
}

.App-modal {
  position: absolute;
  left: -50%; 
  right: -50%;
  width: 15rem;
  top: 5rem;
  @media screen and (min-width: 400px) {
      min-width: 20rem;
  }
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  padding: 1.5rem;
  background-color: $light-gray;
  box-shadow: $box-shadow;
  border-radius: 1rem;
  z-index: 3;

  animation-name: expandUp;
  -webkit-animation-name: expandUp;	

  animation-duration: 1s;	
  -webkit-animation-duration: 1s;

  animation-timing-function: ease;	
  -webkit-animation-timing-function: ease;		

  visibility: visible !important;	
}

.App-footer-audio, .App-footer {
  text-align: center;
  padding-bottom: 1.5rem;
}

.App-footer-audio {
  padding-bottom: 7rem;
}

.listenNotesLogo {
  max-width: 15rem;
}

.logo {
  width: 15rem;
  height: 15rem;
}

@media screen and (min-width: 400px) {
  .logo {
    width: 10rem;
    height: 10rem;
    top: 0;
    position: absolute;
  }
}

body {
  font-family: 'Cairo', sans-serif;
  margin: 0;
  padding: 0;
}

a {
  color: black;
}

button {
  font-family: 'Cairo';
  background-color: $medium-gray;
}

input {
  font-family: 'Cairo', sans-serif;
}

p {
  margin: 0;
}

button:hover:not([disabled]) {
  background-color: $gray;
  cursor: pointer;
}

//loaders
.loader, .loaderSmall, .loaderTiny {
  border-radius: 100%;
  animation: spin 1.8s linear infinite;
}

.loader {
  border: 16px solid $light-gray;
  border-top: 16px solid $blue;
  width: 120px;
  height: 120px;
  margin: 0 auto;
}

.loaderSmall {
  border: 16px solid $light-gray;
  border-top: 16px solid $blue;
	height: 3rem;
	width: 3rem;
}

.loaderTiny {
  border: 5px solid $gray;
  border-top: 5px solid $blue;
	height: 1rem;
	width: 1rem;
}

@keyframes spin {
  0% { transform: rotate(0deg); }
  100% { transform: rotate(360deg); }
}

@keyframes expandUp {
  0% {
      transform: translateY(100%) scale(0.6) scaleY(0.5);
  }
  60%{
      transform: translateY(-7%) scaleY(1.12);
  }
  75%{
      transform: translateY(3%);
  }	
  100% {
      transform: translateY(0%) scale(1) scaleY(1);
  }	
}

@-webkit-keyframes expandUp {
  0% {
      -webkit-transform: translateY(100%) scale(0.6) scaleY(0.5);
  }
  60%{
      -webkit-transform: translateY(-7%) scaleY(1.12);
  }
  75%{
      -webkit-transform: translateY(3%);
  }	
  100% {
      -webkit-transform: translateY(0%) scale(1) scaleY(1);
  }	
}