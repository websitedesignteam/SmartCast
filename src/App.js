import React, { useState, useRef, useEffect } from 'react';
import './App.scss';
import Navbar from './component/Navbar/Navbar';
import Auth from './component/Auth/Auth';
import Search from './component/Search/Search';
import Podcast from './page/Podcast/Podcast';
import Episode from './page/Episode/Episode';
import Home from './page/Home/Home';
import Landing from './page/Landing/Landing'
import Genres from './page/Genres/Genres';
import SearchPage from './page/Search/SearchPage';
import UserProfile from './page/Profile/UserProfile'
import {
  BrowserRouter as Router,
  Route,
  Switch,
	Redirect,
} from "react-router-dom";
import { withSearchContext } from "state/Search/withSearchContext";
import AudioFooter from './component/AudioFooter/audioFooter';
import { useIsActive, useOnClickOutside } from 'hooks';
import { getUser, getTokenValidation, getNewToken } from "./utils/api";
import { baseUrl, errorSessionExpired } from 'utils/constants';
import { EpisodeCard } from 'component/Podcast';

function App() {
  //global app states
  const [isLoading, setIsLoading] = useState(true);
  const [appHeight, setAppHeight] = useState(null);
  const [audio, setAudio] = useState(() => JSON.parse(localStorage.getItem('audio')) || {}); // audio = {podcastName, episodeName, podcastPublisher, audio}
  const [user, setUser] = useState(() => JSON.parse(localStorage.getItem('user')) || {}); //user = {username, authToken, refreshToken}
  const authModalState = useIsActive();
  const stayLoggedIn = useIsActive();

  //audioplayer
  const openAudioPlayer = (newAudio) => {
    setAudio(newAudio);
    localStorage.setItem("audio", JSON.stringify(newAudio));
  }

  const closeAudioPlayer = () => {
    setAudio({});
    localStorage.removeItem("audio");
  }

  //auth
  const getUserAPI = (currentUser) => {
    if (!currentUser) currentUser = user;
    const { access_token } = currentUser;
    getUser({access_token})
    .then((response) => {
      closeAuthModal();
      const userData = response.data;
      const allUserData = { 
        ...currentUser,
        ...userData, 
      }
      localStorage.setItem("user", JSON.stringify(allUserData));
      setUser(allUserData);
    })
    .catch((error) => {
      console.log(error);
    });
  }

  const logoutUser = () => {
    setUser({});
    localStorage.removeItem("user");
  }

  const validateToken = () => {
    const { access_token, refresh_token, username } = user;
    if (!!access_token && !!refresh_token && !!username) {
      getTokenValidation({access_token})
        .catch(() => {
          if (stayLoggedIn.isActive) {
            const data = {
              refresh_token,
              username,
            };

            //if user wants to stay logged in for 30 days
            getNewToken(data)
            .then((response) => {
                const allUserData = {
                  ...user,
                  ...response.data,
                }
                localStorage.setItem("user", JSON.stringify(allUserData));
                setUser(allUserData);
            })
            .catch((error)=> {
              console.log(error);
              setUser({});
              stayLoggedIn.deactivate();
              localStorage.removeItem("user");
            })
          } else {
            alert(errorSessionExpired);
            setUser({});
            localStorage.removeItem("user");
          }
        })
    }
  }

  //modal
  const openAuthModal = () => {
    authModalState.activate();
    // document.body.style.position = 'fixed';
    // document.body.style.overflowY = "scroll";
  }

  const closeAuthModal = () => {
    authModalState.deactivate();
    // document.body.style.position = 'static';
    // document.body.style.overflowY = 'auto';
  }

  //refs
  const authModalRef = useRef();
  useOnClickOutside(authModalRef, closeAuthModal);
  const appRef = useRef();

  const getAppHeight = () => {
    const el = appRef.current;
    if (!el) return
    const currentAppHeight = el.scrollHeight;
    setAppHeight(currentAppHeight);
  }

  useEffect(()=> {
    getAppHeight();
  }, [])



  return (
    <div className="App">
      <Router>
        <div className="App-content" ref={appRef}>  
          {(authModalState.isActive && !user.access_token) && 
            <>
            <div className="App-modal" ref={authModalRef}>
              <Auth loginUser={getUserAPI} type="login" onSuccessVerification={closeAuthModal} stayLoggedIn={stayLoggedIn} /> 
            </div>
            <div className="App-blur" style={{height: appHeight}} />
            </>
          }

          <Switch>
          <Route exact path="/home"
              render={routeProps =>
              <>
              <Navbar logoutUser={logoutUser} openAuthModal={openAuthModal} user={user}/>
        
        {audio.audioUrl && 
            <AudioFooter audio={audio} closeAudioPlayer={closeAudioPlayer} />}
                <Search {...routeProps} /> 
                <Home {...routeProps} />
              </>
            }/>

            <Route exact path="/" render={routeProps => 
              <>
                <Redirect {...routeProps} to="/welcome" /> 
              </>
            } />
            <Route exact path="/welcome" render={routeProps => 
              <> 
                <Landing {...routeProps}/>
              </>
            } />

            <Route exact path="/auth/:authType"
              render={routeProps =>
              <>
              <Navbar logoutUser={logoutUser} openAuthModal={openAuthModal} user={user}/>
        
        {audio.audioUrl && 
            <AudioFooter audio={audio} closeAudioPlayer={closeAudioPlayer} />}
                {!!user.access_token 
                ? <Redirect {...routeProps} to="/" /> 
                :<div className="App-background">
                  <Auth {...routeProps} />
                </div>}
              </>
              }/>

            <Route exact path="/profile" 
             render={routeProps =>
              <>
              <Navbar logoutUser={logoutUser} openAuthModal={openAuthModal} user={user}/>
        
        {audio.audioUrl && 
            <AudioFooter audio={audio} closeAudioPlayer={closeAudioPlayer} />}
              {!user.access_token ? <Redirect {...routeProps} to="/" /> : <UserProfile {...routeProps} />}
              </>
             }
            />

            <Route exact path="/genres"
              render={routeProps =>
              <>
              <Navbar logoutUser={logoutUser} openAuthModal={openAuthModal} user={user}/>
        
        {audio.audioUrl && 
            <AudioFooter audio={audio} closeAudioPlayer={closeAudioPlayer} />}
                <Search {...routeProps} />
                <Genres {...routeProps} />
              </>
            }/>

            <Route exact path="/genres/:genreName2/:genreName" component={Genres}
              render={routeProps =>
              <>
              <Navbar logoutUser={logoutUser} openAuthModal={openAuthModal} user={user}/>
        
        {audio.audioUrl && 
            <AudioFooter audio={audio} closeAudioPlayer={closeAudioPlayer} />}
              <Search {...routeProps} />
              <Genres {...routeProps} />
              </>
            }/>

            <Route exact path="/search/results/:searchTerm/:searchType"
              render={routeProps =>
              <>
              <Navbar logoutUser={logoutUser} openAuthModal={openAuthModal} user={user}/>
        
        {audio.audioUrl && 
            <AudioFooter audio={audio} closeAudioPlayer={closeAudioPlayer} />}
              <Search {...routeProps}/>
              <SearchPage {...routeProps}/>
              </>
            }/>

            <Route exact path="/podcast/:podcastID/episode/:episodeID" 
              render={routeProps =>
              <>
              <Navbar logoutUser={logoutUser} openAuthModal={openAuthModal} user={user}/>
        
        {audio.audioUrl && 
            <AudioFooter audio={audio} closeAudioPlayer={closeAudioPlayer} />}
                <Search {...routeProps} />
                <Episode {...routeProps} setUser={setUser} openAudioPlayer={openAudioPlayer} user={user} validateToken={validateToken} />
              </>
            }/> 

            <Route exact path="/podcast/:podcastID"
              render={routeProps =>
              <>
              <Navbar logoutUser={logoutUser} openAuthModal={openAuthModal} user={user}/>
        
        {audio.audioUrl && 
            <AudioFooter audio={audio} closeAudioPlayer={closeAudioPlayer} />}
              <Search {...routeProps} />
              <Podcast {...routeProps} user={user} validateToken={validateToken} setUser={setUser} />
              </>
            }/> 
            <Route exact path="/episodeBox" component={EpisodeCard} />
          </Switch>
      </div>
      <div className={audio.audioUrl ? "App-footer-audio" : "App-footer"}>
        <img className="listenNotesLogo" src={baseUrl+"/assets/listen-api-logo.png"} alt="This site is powered by Listen Notes" />
      </div>
  
      </Router>
    </div>
  );
}

export default withSearchContext(App);
