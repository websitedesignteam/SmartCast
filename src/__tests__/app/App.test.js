import React from 'react';
import { shallow } from 'enzyme';
import App from 'src/App';

test("should render the App", () => {
    const component = shallow(<App />);
    // expect(component.getElements()).toMatchSnapshot();
});