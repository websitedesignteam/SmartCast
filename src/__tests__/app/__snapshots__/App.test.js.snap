// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the App 1`] = `
Array [
  <SearchContextProvider>
    <App />
  </SearchContextProvider>,
]
`;
